### PR TITLE
[#3799] Update gems

### PR DESCRIPTION
**Pull request by garethrees**
Work on #3799 

This PR:

- Adds some notes about specifying gem versions (27b9b27) – @crowbot & @stevenday weigh in on this if you like.
- Updates version constraints in line with the above
- Adds `<` constraints where we're blocked from updating (e.g. no Ruby 1.9 support or needs a more recent Rails)
- Updates a few gems
- Doesn't touch bootstrap or foundation gems

I've added some todos on #3799 to cover gem upgrades (minor or major version bumps).

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,78 @@
...
+# This will install the latest version of the 1.2 series. When 1.3 or 2.0 are
+# released, we can assume there may be a breaking change.
+#
+# If we need to be more specific than this, apply one of the following
+# guidelines. Use commit messages to describe why the extra specificity is
+# required. Links to changelogs are very appreciated.
+#
+# If we need greater than a specific PATCH level, specify that:
+#
+#    gem 'foo', '~> 1.2.7'
```
Comment (garethrees) [inline Gemfile]: I'm wondering whether this should be

```ruby
gem 'foo', '~> 1.2.0', '>= 1.2.7'
```

Its the same, but it does make it more explicit that we really do need `1.2.7+`

Comment (stevenday) [inline Gemfile]: Yep, I always have to look up what `~>` actually means

File: Gemfile [hunk 2]
```diff
@@ -70,40 +144,40 @@ gem 'mail', :git => 'https://github.com/mikel/mail', :branch => '2-6-stable'
...
-  gem 'rspec-activemodel-mocks', '~> 1.0.1'
+  gem 'rspec-activemodel-mocks', '~> 1.0.0'
   gem 'rspec-rails', '~> 3.4.0'
-  gem 'pry', '~> 0.10.4'
-  gem 'pry-debugger', '~> 0.2.3', :platforms => :ruby_19
-    gem 'public_suffix', '< 1.5.0'
+  gem 'pry', '~> 0.10.0'
+  gem 'pry-debugger', '~> 0.2.0', :platforms => :ruby_19
+    gem 'public_suffix', '~> 1.4.0', '< 1.5.0'
 end
```
Comment (lizconlan) [inline Gemfile]: I think we can finally add this now:

`gem 'pry-byebug', '~> 3.4.0', :platforms => [:ruby_20, :ruby_21]`

\o/

File: Gemfile [hunk 3]
```diff
@@ -169,7 +169,7 @@ group :test, :development do
   gem 'rspec-rails', '~> 3.4.0'
   gem 'pry', '~> 0.10.0'
   gem 'pry-debugger', '~> 0.2.0', :platforms => :ruby_19
-    gem 'public_suffix', '< 1.5.0'
+    gem 'public_suffix', '~> 1.4.0', '< 1.5.0'
```
Comment (lizconlan) [inline Gemfile]: bonus feature...  I think we can finally add this now:

`gem 'pry-byebug', '~> 3.4.0', :platforms => [:ruby_20, :ruby_21]`

\o/

Comment (garethrees) [inline Gemfile]: nope, bundler we rely on is too old so doesn't support the latest ruby platforms. sigh.

**Pull request merged**